lib/libstorj/src/uploader.c

```diff
@@ -0,0 +1,250 @@
+#include "uploader.h"
+
+static uv_work_t *uv_work_new()
+{
+    uv_work_t *work = malloc(sizeof(uv_work_t));
+    return work;
+}
+
+static void cleanup_upload_work(uv_work_t *work)
+{
+    storj_upload_state_t *state = work->data;
+
+    cleanup_state(state);
+    free(work);
+}
+
+static void cleanup_state(storj_upload_state_t *state)
+{
+    if (state->original_file) {
+        fclose(state->original_file);
+    }
+
+    state->finished_cb(state->error_status, state->info, state->handle);
+
+    free(state);
+}
+
+static void after_get_file_info(uv_work_t *work, int status)
+{
+    if (status) {
+        // TODO: should finished_cb be called here?
+        free(work);
+        return;
+    }
+
+    get_file_info_request_t *req = work->data;
+    uv_work_t *upload_work = req->handle;
+    storj_upload_state_t *state = upload_work->data;
+
+    if (state->error_status) {
+        /* Currently, if status == 0 && state->error_status != 0,
+           finished_cb gets called; this is inconsistent. */
+        goto cleanup;
+    }
+
+    storj_file_meta_t *info = state->info;
+    STORJ_RETURN_SET_STATE_ERROR_IF_LAST_ERROR;
+
+    info->filename = strdup(req->file->filename);
+
+    info->created = strdup(req->file->created);
+    info->mimetype = strdup(req->file->mimetype);
+    info->bucket_id = strdup(req->file->bucket_id);
+    info->id = strdup(req->file->id);
+    info->size = req->file->size;
+
+cleanup:
+    cleanup_upload_work(upload_work);
+    storj_free_get_file_info_request(req);
+    free(work);
+}
+
+static void queue_get_file_info(uv_work_t *work, int status)
+{
+    storj_upload_state_t *state = work->data;
+    STORJ_RETURN_SET_STATE_ERROR_IF_LAST_ERROR;
+    if (state->error_status) {
+        after_get_file_info(work, state->error_status);
+        return;
+    }
+
+    storj_bridge_get_file_info(state->env, state->bucket_id, state->file_name,
+                               strdup(state->encryption_access), work,
+                               after_get_file_info);
+}
+
+static void store_file(uv_work_t *work)
+{
+    storj_upload_state_t *state = work->data;
+
+    size_t buf_len;
+    uint8_t *buf;
+    while (state->uploaded_bytes < state->file_size) {
+        size_t remaining_size = state->file_size - state->uploaded_bytes;
+        if (remaining_size >= state->buffer_size) {
+            buf_len = state->buffer_size;
+        } else {
+            buf_len = remaining_size;
+        }
+
+        buf = malloc(buf_len);
+        size_t read_size = fread(buf, sizeof(char), buf_len, state->original_file);
+        // TODO: what if read_size != buf_len?
```

Comment:
According to the documentation (http://www.cplusplus.com/reference/cstdio/fread/) this can happen either on EOF or error. I think EOF is excluded here because we request reading of no more than the remaining bytes in the file. So, the only case remaining is a read error.